Allow unsmoothed model in SurfaceCut

mikebind opened the pull request:
Adds checkbox to govern whether cutting surface is
smoothed (prior behavior) or unsmoothed (new behavior) and implements the associated
functionality of switching between them

-----------
This is my first pull request to a Slicer project.  I tried to follow the style and patterns already present in the code, but there's probably a higher-than-average chance I've done something that doesn't fit at least the preferred style.  Please edit or request changes as needed!

mikebind committed:
Allow unsmoothed model in SurfaceCut
Adds checkbox to govern whether cutting surface is
smoothed (prior behavior) or unsmoothed (new behavior) and implements the associated
functionality of switching between them

lassoan commented on SegmentEditorSurfaceCut/SegmentEditorSurfaceCutLib/SegmentEditorEffect.py:
```diff
@@ -175,6 +192,17 @@ def onOperationSelectionChanged(self, operationName, toggle):
     self.scriptedEffect.setParameter("Operation", operationName)
 
+  def onSmoothModelCheckboxStateChanged(self, newState):
+    logging.info('onSmoothModelCheckboxStateChanged, new state is: %s'%(str(newState)))
```
no need to log this, but if you really want to then change it to debug level (info message is displayed in the Python console, which would be too much noise)

mikebind replied:
Thanks, that was a debugging message, most of which I had put in as logging.debug, and I just missed this one when I was removing those.  I'll take this one out also

lassoan replied:
Instead of adding temporary log messages, debugging is much easier and faster if you connect a debugger to Slicer (Visual Studio Code, PyCharm, VIsual Studio, and Eclipse are all supported). See more information here: https://www.slicer.org/wiki/Documentation/Nightly/Extensions/DebuggingTools

lassoan commented on SegmentEditorSurfaceCut/SegmentEditorSurfaceCutLib/SegmentEditorEffect.py:
```diff
@@ -175,6 +192,17 @@ def onOperationSelectionChanged(self, operationName, toggle):
+  def onSmoothModelCheckboxStateChanged(self, newState):
+    logging.info('onSmoothModelCheckboxStateChanged, new state is: %s'%(str(newState)))
+    if newState==qt.Qt.Checked:
+      self.scriptedEffect.setParameter("SmoothModel", 'True')
```
This is correct, but for consistency with how we save Boolean values, save it as 0/1 instead see this example:

https://github.com/Slicer/Slicer/blob/62db2030cba19974d8e0cd556c084a9e00049f4f/Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorLogicalEffect.py#L151-L152

lassoan commented on SegmentEditorSurfaceCut/SegmentEditorSurfaceCutLib/SegmentEditorEffect.py:
```diff
@@ -377,7 +405,8 @@ def onSegmentEditorNodeModified(self, observer, eventid):
     if not self.segmentMarkupNode or not self.segmentModel:
       return
-    self.logic.updateModelFromMarkup(self.segmentMarkupNode, self.segmentModel)
+    smoothModelFlag = self.smoothModelCheckbox.isChecked()
```
Since this method is kind of "logic" (not GUI), it is nicer if we don't refer to GUI widgets but get the smoothModel flag from the effect parameter instead.

lassoan commented on SegmentEditorSurfaceCut/SegmentEditorSurfaceCutLib/SegmentEditorEffect.py:
```diff
@@ -165,6 +175,13 @@ def updateGUIFromMRML(self):
       operationButton = list(self.buttonToOperationNameMap.keys())[list(self.buttonToOperationNameMap.values()).index(operationName)]
       operationButton.setChecked(True)
+    
+    if self.scriptedEffect.parameter("SmoothModel") == 'True':
```
This is correct, but for consistency with how other effects store Boolean values, please use 0/1 instead. See example here:

https://github.com/Slicer/Slicer/blob/62db2030cba19974d8e0cd556c084a9e00049f4f/Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorLogicalEffect.py#L141-L144

lassoan commented on SegmentEditorSurfaceCut/SegmentEditorSurfaceCutLib/SegmentEditorEffect.py:
```diff
@@ -94,6 +94,14 @@ def setupOptionsFrame(self):
+    self.smoothModelCheckbox.setChecked(True) # model smoothing initial default is True
+    self.smoothModelCheckbox.setToolTip("Model is smoothed if checked, faceted if unchecked")
+    smoothModelLayout.addWidget(self.smoothModelCheckbox, 0, 0)
+    self.scriptedEffect.addLabeledOptionsWidget("Smooth model:", smoothModelLayout)
```
If you only need to display a single widget then it should not be necessary to put it into a layout.
Have you tried to just use the checkbox directly in addLabeledOptionsWidget?

mikebind replied:
Yes, that worked, thanks

lassoan reviewed (commented):
Thanks a lot for your contribution! This works very nicely. I just added a few comments to improve style&consistency of the code.

mikebind committed:
Adjusts style to be more consistent
Checkbox state reflected by 0/1 parameter
Lone widget item not placed in extra layer of layout
Logic-like method altered to use parameter node instead of widget state

mikebind commented:
I think this addresses all the comments.  Let me know if you see anything else.  I did notice that the module occasionally generates the following warning as the model is updated:  
`Generic Warning: In D:\D\P\S-0-E-b\MarkupsToModel\MarkupsToModel\Logic\vtkSlicerMarkupsToModelClosedSurfaceGeneration.cxx, line 311
Extent ranges not provided in order largest to smallest. Unexpected results may occur.
`
However, upon further checking, the module sometimes generates this warning without any of the new changes, and it doesn't appear to be related to any of the new code.

lassoan commented:
Perfect, this works very well and nicely implemented. Thank you for your contribution!

I've squashed the commits and pushed here: https://github.com/lassoan/SlicerSegmentEditorExtraEffects/commit/759d014efcb90bbfa18c7e091c97d039f8a06ae5
